internal/quickfort/dig.lua

```diff
@@ -457,52 +457,115 @@ local function do_traffic_restricted(ctx)
-    h=do_channel,
-    u=do_up_stair,
-    j=do_down_stair,
-    i=do_up_down_stair,
-    r=do_ramp,
-    z=do_remove_ramps,
-    t=do_mine,
-    p=do_gather,
-    s=do_smooth,
-    e=do_engrave,
-    F=do_fortification,
-    T=do_track,
-    v=do_toggle_engravings,
-    M=do_toggle_marker,
-    n=do_remove_construction,
-    x=do_remove_designation,
-    bc=do_claim,
-    bf=do_forbid,
-    bm=do_melt,
-    bM=do_remove_melt,
-    bd=do_dump,
-    bD=do_remove_dump,
-    bh=do_hide,
-    bH=do_unhide,
-    oh=do_traffic_high,
-    on=do_traffic_normal,
-    ol=do_traffic_low,
-    ['or']=do_traffic_restricted,
+local dig_db = {
+    d={action=do_mine, use_priority=true},
+    h={action=do_channel, use_priority=true},
+    u={action=do_up_stair, use_priority=true},
+    j={action=do_down_stair, use_priority=true},
+    i={action=do_up_down_stair, use_priority=true},
+    r={action=do_ramp, use_priority=true},
+    z={action=do_remove_ramps, use_priority=true},
+    t={action=do_mine, use_priority=true},
+    p={action=do_gather, use_priority=true},
+    s={action=do_smooth, use_priority=true},
+    e={action=do_engrave, use_priority=true},
+    F={action=do_fortification, use_priority=true},
+    T={action=do_track, use_priority=true},
+    v={action=do_toggle_engravings},
+    -- the semantics are unclear if the code is M but m or force_marker_mode is
+    -- also specified. skipping all other marker mode settings when toggling
+    -- marker mode seems to make the most sense.
+    M={action=do_toggle_marker, skip_marker_mode=true},
+    n={action=do_remove_construction, use_priority=true},
+    x={action=do_remove_designation},
+    bc={action=do_claim},
+    bf={action=do_forbid},
+    bm={action=do_melt},
+    bM={action=do_remove_melt},
+    bd={action=do_dump},
+    bD={action=do_remove_dump},
+    bh={action=do_hide},
+    bH={action=do_unhide},
+    oh={action=do_traffic_high},
+    on={action=do_traffic_normal},
+    ol={action=do_traffic_low},
+    ['or']={action=do_traffic_restricted},
 }
 
-local function dig_tile(ctx, code, marker_mode)
+-- handles marker mode 'm' prefix and priority suffix
+local function extended_parser(_, keys)
+    local marker_mode = false
+    if keys:startswith('m') then
+        keys = string.sub(keys, 2)
+        marker_mode = true
+    end
+    local found, _, code, priority = keys:find('^(%D*)(%d*)$')
+    if not found then return nil end
+    if #priority == 0 then
+        priority = 4
+    else
+        priority = tonumber(priority)
+        if priority < 1 or priority > 7 then
+            log('priority must be between 1 and 7 (inclusive)')
+            return nil
+        end
+    end
+    if #code == 0 then code = 'd' end
+    if not rawget(dig_db, code) then return nil end
+    local custom_designate = copyall(dig_db[code])
+    custom_designate.marker_mode = marker_mode
+    custom_designate.priority = priority
+    return custom_designate
+end
+
+setmetatable(dig_db, {__index=extended_parser})
```

Comment:
Ah, extended_parser will never get called for keys that are already in `dig_db` (like "d" from my examples).
I'm not sure if you accounted for this - it looks like extended_parser is intended to handle keys without numbers, which I don't think it can be called with.

Update: I see set_priority() guarded by a check for dig_db[key].use_priority, but it doesn't seem to be called when the priority isn't specified, and it probably should be called with 4. I'm assuming you have a better idea of how to address this.

Reply:
it still gets called when there is an 'm' prefix for marker mode, so "md" would get here (and assign the default priority of 4).

I wasn't calling set_priority() when priority isn't specified since the game interprets lack of an event block as all priorities == 4, and I didn't think I needed to mess with them. but in this case what *might* be happening is an event block already exists with a 0 value and I'm not overwriting it? It's true that I hadn't considered that case. it could happen with non-zero values as well, I think. if a previous designation was at priority N, was cleared (which does not clear the event block, IIRC), and then a blueprint without an explicit priority was applied. I think I can fix this by assigning priority=4 to the db records.

Reply:
Yep, block events cover the whole block. You're not the first to run into issues caused by this. Looks good now!